Add legacy taiko swell (spinner)

mcendu opened the pull request:
Fixes #13601. Based on/closes #21601.

> I tried to structure this so that there's a clear divide between gameplay logic and visuals to keep `DrawableSwell` from getting too big. `DrawableSwell` contains all of the gameplay logic and creates two `SkinnableDrawable`: the swell circle piece and the swell body. All of the individual pieces that make up the swell body are then constructed or retrieved in `DefaultSwell` / `LegacySwell`, where all visual animations/fades/changes to the swell (both the swell body and swell circle piece) also happen.
> 
> Did my best to make this look as close to stable as I could, but discrepancies are possible.

https://github.com/user-attachments/assets/200ba15d-1f96-4834-b673-24f8eac20f7c

Joppe27 committed:
Add legacy taiko swell

Joppe27 committed:
Adjust existing test to accommodate swell size

Joppe27 committed:
Add test for expire delay

Delaying the expiry of the drawable hitobject can potentially be dangerous and gameplay-altering when user inputs are accidentally handled. This is why I found a test necessary.

Joppe27 committed:
Rename methods and skin component + add comments

mcendu committed:
Merge taiko swell components

Per <https://github.com/ppy/osu/pull/21601/files#r1064167756>, taking a
variation of the "Make all swell main pieces implement
ISkinnableSwellPart" path. Should clean the interface up enough for
further refactors.

mcendu committed:
remove ISkinnableSwell

This commit removes ISkinnableSwell for taiko swell animations. In place
of it, an event named UpdateHitProgress is added to DrawableSwell, and
the skin swells are converted to listen to said event and
ApplyCustomUpdateState, like how spinner skinning is implemented for
std.

mcendu committed:
remove spinner-osu.png workaround

https://github.com/ppy/osu/issues/22084

mcendu committed:
change legacy spinner animations to match stable

Also removed a few fallbacks pointed out in code review that I don't
understand.

mcendu committed:
fix code style

bdach commented:
Visual comparisons would go a long way to ease review. Although I guess end-of-day I will still have to do a stable cross-check so I dunno. Would be appreciated but I'll probably do them myself anyway.

Joppe27 commented:
Thanks for reviving my PR! I played around with this a bit and saw something which seems off: 

https://github.com/user-attachments/assets/ac0b3213-dbd8-43e9-933e-dd9be8909a80

This is `clearAnimation`, both the location and duration seem incorrect to me. I recall that when I attempted to implement bdach's feedback I also was not able to just copy stable's values and be done with it (which is part of why I wasn't able to complete the PR, too much time to investigate all of the stable-lazer discrepancies).

---

Another issue I had trouble with which probably needs to be addressed is the hardcoded `LifetimeEnd` extension, which was [already pointed out by bdach](https://github.com/ppy/osu/pull/21601#discussion_r1064167942) in the original PR. I'll explain below what I learned during my attempt at fixing this, in the hope it saves you some time. 

The following happens when completing a swell:
1. Player completes swell
2. `ApplyMaxResult()` in `DrawableSwell`
3. `UpdateState(ArmedState.Hit)` in `DrawableHitObject`
4. `UpdateHitStateTransforms(ArmedState.Hit)` in `DrawableHitObject`, can set `DrawableSwell`'s lifetime end in this function's override in `DrawableSwell`
5. `DrawableSwell` expires!
6. *Only then* `ApplyCustomUpdateState` gets invoked with `ArmedState.Hit`, too late to play the clear animation as `DrawableSwell` already expired

In the original review comment it was suggested to use `LatestTranformEndTime` in the override (4) to postpone `DrawableSwell`'s expiry until the clear animation fully completes, but this is not possible as the transforms only get applied after `ApplyCustomUpdateState` gets invoked (6) - the transforms don't exist yet at `UpdateHitStateTransforms` (4).

With your approach you could create another event which gets invoked *in* (instead of after) `UpdateHitStateTransforms` that applies the necessary transforms for the clear animation, after which you can extend `DrawableSwell`'s life using `LatestTransformEndTime` as above. You could also accomplish this in `animateSwellProgress` without creating a new event, by applying the transforms when `requiredHits - numHits` is 0 (although this would be before the DHO is officially considered "hit" so maybe less ideal). There are likely better ways, just giving you some ideas!

It's worth noting that this is also "hardcoded" for osu!standard in `DrawableSpinner`, but the transforms in `LegacySpinner` take the hitobject duration (which includes the delay) into account which works just as well. This is another way to fix the issue which might even be simpler. The clear animation for legacy taiko skins is much longer, but this is no problem: `HandleUserInput = false` and the test from the original PR make sure this doesn't affect gameplay. You would then subtract from the hitobject duration (which at `ArmedState.Hit` is equal to the hardcoded delay) the value necessary to end up with the correct transform length for the legacy, default and argon swell. This code might be a bit less transparent though but since it's been done before probably fine.

mcendu committed:
further adjust swell behavior

The outstanding visual issues of the clear animation is fixed. The
HandleUserInput state management is removed as it no longer seems
necessary.

mcendu commented on osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableSwell.cs:
```diff
@@ -248,28 +156,23 @@ protected override void CheckForResult(bool userTriggered, double timeOffset)
                 case ArmedState.Idle:
-                    expandingRing.FadeTo(0);
+                    // Only for rewind support. Reallows user inputs if swell is rewound from being hit/missed to being idle.
+                    HandleUserInput = true;
```
Is this state management still necessary? I did a few plays on my build with this line removed and everything seems to work normally.

Joppe27 replied:
No, this isn't necessary anymore. This was apparently fixed separately after I created the original PR (see https://github.com/ppy/osu/pull/24196):

https://github.com/ppy/osu/blob/e794389fe83644323a563a343338e282783b53b1/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableSwell.cs#L204-L205

This also means the test I added back then has become redundant, you can remove it from this PR.

mcendu committed:
remove skinning section swell delay test

Replaced by TestHitSwellThenHitHit in TestSceneSwellJudgements.

mcendu committed:
code cleanup

bdach commented:
I should really not be surprised at this point but visually this is not even close.

https://github.com/user-attachments/assets/23c6fb06-3d15-4714-b4eb-e09a0dcbc004

I guess I'm gonna have to try and get it close myself, as usual.

bdach committed:
Fix remaining hit counter scaling in the incorrect direction

bdach committed:
Adjust things to be closer to stable (but not close enough yet)

bdach commented:
Still not done with this.

Things that are fixed:

- Remaining hit counter scales in correct direction
- Some incorrect easing, duration, and sizing specs changed

To be fixed still:

- The entirety of the clear animation looks wrong. Not positioned right, not moving right, not scaling right.
- On stable the entirety of the swell is like slightly transparent. Haven't tracked why it's doing that yet.
- `animateSwellProgress()` is incorrect because any transforms applied therein will just get reverted when the DHO goes through the proper `UpdateState()` flow.

peppy commented:
> Stable also does this thing wherein it's just positioned dead centre of game window but only on 4:3 aspect ratio because of stable's completely wonky widescreen handling HELP ANYONE HELP PLEASE---

I don't think this needs to be replicated. As long as it roughly matches it's fine (even if it's different on widescreen vs 4:3).

The stable code [I'm guessing you're looking at](https://github.com/peppy/osu-stable-reference/blob/3ea48705eb67172c430371dcfc8a16a002ed0d3d/osu!/GameplayElements/HitObjects/Taiko/SpinnerTaiko.cs#L73-L74) is definitely not worth trying to piece apart. Not for legacy taiko skins.

> On stable the entirety of the swell is like slightly transparent.

It seems like the maximum opacity is always 80%, applied in a [weird way on stable](https://github.com/peppy/osu-stable-reference/blob/3ea48705eb67172c430371dcfc8a16a002ed0d3d/osu!/GameplayElements/HitObjects/Taiko/SpinnerTaiko.cs#L131-L135). Fine to match that in the initial transform.

Let me know if you need any more feedback specifically on the stable part / animations parts. I don't think this needs to match 1:1, as long as it doesn't get in the player's way more than it does on stable.

bdach committed:
Adjust things further

bdach committed:
Fix transforms from swell progress being cleared on completion by not using transforms

bdach committed:
Add missing animation of warning sprite

bdach committed:
Fix clear sample potentially playing multiple times simultaneously

bdach commented:
I think I'm done with this branch.

https://github.com/user-attachments/assets/406c8fbf-04bb-4ccf-9cc6-8011ea3dfc4a

In the video above stable on left, lazer on right. Note that for each skin the sync point of both clients is not the start of the swell but the start of the clear animation for easier convenience of comparing that.

Kinda baffling how much work this took to get anywhere close. And don't get it twisted, I only leaned on source code for a few things. Most of alignment was visual ballpark.

Requesting second review from someone else. Code-wise probably the most suspect part is https://github.com/ppy/osu/pull/31746/commits/0ac08158e33867092f76f94b1534ba3bc1ce962c. `halfTime` values in the `Interpolation.DampContinuously()` calls there were calculated by taking the original easings and ballparking how far along the original easing reached the value of 0.5.

peppy reviewed (approved):
ship-it🚢
